File: PrimeiraAplicacao.Funcional/Funcoes.Funcional/FuncaoAltaOrdem.fs
module FuncaoAltaOrdem

let executarAcaoSobreElementos lista acao =
    lista |> List.iter acao

let imprimirNomes nomes = executarAcaoSobreElementos nomes (printfn "Olá %s")
 
let imprimirDobroDeUmNumero numero = printfn "%i." (numero * 2)

let imprimirDobroDosNumeros numeros =  executarAcaoSobreElementos numeros (fun numero -> printfn "%i." (numero *2))

let multiplicarEImprimirNumero multiplicador numero = printfn "%i." (numero * multiplicador)

let imprimirListaDeNumeros multiplicador numeros =
    executarAcaoSobreElementos numeros
        (multiplicarEImprimirNumero multiplicador)

let multiplicarEImprimirNumeroCurrying multiplicador = 
    let imprimirNumero numero =
        printfn "%i." (numero * multiplicador)
    
    imprimirNumero

File: PrimeiraAplicacao.Funcional/Funcoes/FuncaoAltaOrdem.cs
using System;
using System.Collections.Generic;

namespace Funcoes
{
    public class FuncaoAltaOrdem
    {
        public static void ExecutarAcaoSobreElementos<TipoElemento>(
            List<TipoElemento> lista, 
            Action<TipoElemento> acao)
        {
            lista.ForEach(acao);
        }

        public static void ImprimirNomes(List<string> nomes)
        {
            ExecutarAcaoSobreElementos(nomes, Console.WriteLine);
        }

        public static void ImprimirDobroDosNumeros(List<int> numeros)
        {
            ExecutarAcaoSobreElementos(numeros, numero => Console.WriteLine(numero * 2));
        }
    }
}
